packages/admin-ui-react/src/system.tsx

```diff
@@ -0,0 +1,254 @@
+      MergeLeft<MergeLeft<{ as: As }, Props>, React.ComponentPropsWithRef<As>>,
+      ComponentStyleProps
+    >
+  ): JSX.Element
+  displayName?: string
+  [__as]: Type
+}
+
+interface AdminUIComponent<Type, Props extends {}>
+  extends React.ForwardRefExoticComponent<
+    Join<
+      MergeLeft<
+        Props,
+        Omit<React.ComponentPropsWithRef<ComponentInfer<Type>>, 'as'>
+      >,
+      ComponentStyleProps
+    >
+  > {
+  /**
+   * Handle a component type
+   * @example
+   * const Button = jsx('button')()
+   *
+   * <Button to="" /> // 🚨 type error
+   * <Button as={GatsbyLink} to="" /> // ✅ all good
+   */
+  <
+    As extends React.ComponentType = Type extends React.ComponentType
+      ? Type
+      : never
+  >(
+    props: Join<
+      MergeLeft<MergeLeft<{ as?: As }, Props>, React.ComponentPropsWithRef<As>>,
+      ComponentStyleProps
+    >
+  ): JSX.Element
+
+  /**
+   * Prioritize elements over components
+   * @example
+   * const Button = jsx('button')()
+   *
+   * <Button href="" /> // 🚨 type error
+   * <Button as="a" href="" /> // ✅ all good
+   */
+  <
+    El extends IntrinsicElementsKeys = Type extends IntrinsicElementsKeys
+      ? Type
+      : never
+  >(
+    props: Join<
+      MergeLeft<MergeLeft<{ as: El }, Props>, React.ComponentPropsWithRef<El>>,
+      ComponentStyleProps
+    >
+  ): JSX.Element
+  displayName?: string
+  [__as]: Type
+}
+
+type Component<
+  T extends React.ElementType,
+  Props extends {} = {}
+> = T extends string
+  ? AdminUIElement<T, Props>
+  : T extends {
+      [__as]: infer TT
+    }
+  ? AdminUIElement<TT, Props>
+  : AdminUIComponent<T, Props>
+
+/**
+ * Create a component typed with the as prop & admin-ui styles
+ * @example
+ * type Props = {}
+ * const Component = createComponent<'element or component', Props>(props => {
+ *  return useElement('element or component', props)
+ * })
+ */
+export function createComponent<
+  As extends React.ElementType,
+  P extends {} = {}
+>(
+  render: (props: PrivateProps<As, P>) => JSX.Element | null
+): Component<As, P> {
+  const AdminUIComponent = (props: any, ref: React.Ref<any>) =>
+    render({ ref, ...props })
+
+  return forwardRef(AdminUIComponent) as unknown as Component<As, P>
+}
+
+type Hook<As extends React.ElementType, P extends {}> = {
+  /**
+   * Describe the hook behavior
+   */
+  (props: PrivateProps<As, P>): PublicProps<As>
+}
+
+/**
+ * Create a hook to hold a component behavior
+ * @example
+ * type Props = {}
+ * const useComponent = createHook<'element or component', Props>(props => props)
+ */
+export function createHook<As extends React.ElementType, P extends {} = {}>(
+  useProps: (props: PrivateProps<As, P>) => PublicProps<As>
+): Hook<As, P> {
+  const useComponent: Hook<As, P> = (props) => {
+    const htmlProps = useProps(props)
+    const copy = {} as typeof htmlProps
+
+    for (const prop in htmlProps) {
+      if (hasOwnProperty(htmlProps, prop) && htmlProps[prop] !== undefined) {
```

Comment:
**Suggestion:** I think if you use the optional chaining wouldn't be necessary to use the function `hasOwnProperty`.
```suggestion
      if (htmlProps?.[prop] !== undefined) {
```